File: src/brs/unconfirmedtransactions/UnconfirmedTransactionStore.java
package brs.unconfirmedtransactions;

import brs.SignumException;
import brs.Transaction;
import brs.peer.Peer;

import java.util.List;

public interface UnconfirmedTransactionStore {

  boolean put(Transaction transaction, Peer peer) throws SignumException.ValidationException;

  Transaction get(Long transactionId);

  boolean exists(Long transactionId);
  
  long getFreeSlot(int numberOfBlocks);

  List<Transaction> getAll();

  List<Transaction> getAllFor(Peer peer);

  void remove(Transaction transaction);

  void clear();

  void resetAccountBalances();

  void markFingerPrintsOf(Peer peer, List<Transaction> transactions);

  void removeForgedTransactions(List<Transaction> transactions);

  int getAmount();
}

File: src/main/java/brs/unconfirmedtransactions/UnconfirmedTransactionStore.kt
package brs.unconfirmedtransactions

import brs.Transaction
import brs.peer.Peer

interface UnconfirmedTransactionStore {
    val all: List<Transaction>

    val amount: Int

    suspend fun put(transaction: Transaction, peer: Peer?): Boolean

    suspend fun get(transactionId: Long?): Transaction?

    suspend fun exists(transactionId: Long?): Boolean

    suspend fun getAllFor(peer: Peer): Collection<Transaction>

    suspend fun remove(transaction: Transaction)

    suspend fun clear()

    /**
     * Review which transactions are still eligible to stay
     * @return The list of removed transactions
     * TODO rename method
     */
    suspend fun resetAccountBalances()

    suspend fun markFingerPrintsOf(peer: Peer?, transactions: Collection<Transaction>)

    suspend fun removeForgedTransactions(transactions: Collection<Transaction>)
}
